cmd/registry-migrator/main.go

```diff
@@ -0,0 +1,212 @@
+	flag.StringVar(&keyFile, "key", "", "path to etcd client certificate key file")
+	flag.StringVar(&endpoints, "endpoints", "http://localhost:2379", "comma-separated list of etcd endpoints")
+	flag.BoolVar(&force, "force", false, "forcefully override existing keys")
+	flag.BoolVar(&delete, "delete", false, "deletes migrated keys")
+	flag.StringVar(&backup, "backup-file", "", "file to store backup to")
+
+	flag.Parse()
+
+	clientv3.SetLogger(grpclog.NewLoggerV2(os.Stderr, os.Stderr, os.Stderr))
+
+	if oldRegistryPrefix == "" {
+		log.Fatalln("registry-prefix cannot be empty")
+	}
+
+	if newRegistryPrefix == "" {
+		log.Fatalln("new-refistry-prefix cannot be empty")
+	}
+
+	if backup == "" {
+		log.Fatalln("backup-file cannot be empty")
+	}
+
+	eps := strings.Split(endpoints, ",")
+	if len(eps) == 0 {
+		log.Fatalln("at least one endpoint must be provided")
+	}
+
+	cc := clientv3.Config{
+		Endpoints:   eps,
+		DialTimeout: dialTimeout,
+	}
+
+	switch {
+	case caFile != "" && certFile != "" && keyFile != "":
+		tlsInfo := transport.TLSInfo{
+			TrustedCAFile: caFile,
+			CertFile:      certFile,
+			KeyFile:       keyFile,
+		}
+
+		tls, err := tlsInfo.ClientConfig()
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		cc.TLS = tls
+	case caFile == "" && certFile == "" && keyFile == "":
+	default:
+		log.Fatalf("--ca, --cert and --key must all be specified when one is provided")
+	}
+
+	cli, err := clientv3.New(cc)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	defer cli.Close()
+
+	lg, err := zap.NewProduction()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	dummy, err := cli.Get(context.TODO(), "/dummy-foo-bar", clientv3.WithKeysOnly())
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	log.Printf("compacting to revision %d", dummy.Header.Revision)
+
+	_, err = cli.Compact(context.TODO(), dummy.Header.Revision, clientv3.WithCompactPhysical())
+	if err != nil {
+		if err == rpctypes.ErrCompacted {
+			log.Println("etcd is already compacted to the latest revision")
+		} else {
+			log.Fatal(err)
+		}
+	}
+
+	log.Printf("compaction successful")
+
+	sp := snapshot.NewV3(lg)
+
+	err = sp.Save(context.TODO(), cc, backup)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	old, err := cli.Get(context.TODO(), oldRegistryPrefix,
+		clientv3.WithPrefix(),
+		clientv3.WithKeysOnly(),
+	)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	log.Printf("Starting from revision %d", old.Header.Revision)
+
+	for _, k := range old.Kvs {
+		key := string(k.Key)
+		newKey := newRegistryPrefix + strings.TrimPrefix(key, oldRegistryPrefix)
+		log.Printf("migrating %q => %q", key, newKey)
+
+		out, err := cli.Get(context.TODO(), key)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		if len(out.Kvs) != 1 {
+			log.Fatalf("key %q is missing", key)
+		}
+
+		compares := []clientv3.Cmp{clientv3.Compare(clientv3.ModRevision(key), "=", out.Kvs[0].ModRevision)}
+		if !force {
+			compares = append(compares, clientv3util.KeyMissing(newKey))
+		}
+
+		txn, err := cli.Txn(context.TODO()).
+			If(compares...).
```

Comment:
If one of the key fails the comparison, this algorithm will skip migrating only that key and continue with others.  Is this expected goal?

Reply:
No. `log.Fatal` exits and it'll stop the tool.